<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
<meta name="viewport"
      content="width=device-width, initial-scale=1.0, maximum-scale=1.0, minimum-scale=1.0">
<meta http-equiv="X-UA-Compatible" content="ie=edge">

    <meta name="author" content="Liang Xu">


    <meta name="subtitle" content="This is Liang's blog for life and work.">




<title>Get genomes, annotate genomes and analyze genomes  | Liang Xu</title>



    <link rel="icon" href="/favicon.ico">




    <!-- stylesheets list from _config.yml -->
    
    <link rel="stylesheet" href="/css/style.css">
    



    <!-- scripts list from _config.yml -->
    
    <script src="/js/script.js"></script>
    
    <script src="/js/tocbot.min.js"></script>
    



    
    
        <!-- MathJax配置，可通过单美元符号书写行内公式等 -->
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
    "HTML-CSS": {
        preferredFont: "TeX",
        availableFonts: ["STIX","TeX"],
        linebreaks: { automatic:true },
        EqnChunk: (MathJax.Hub.Browser.isMobile ? 10 : 50)
    },
    tex2jax: {
        inlineMath: [ ["$", "$"], ["\\(","\\)"] ],
        processEscapes: true,
        ignoreClass: "tex2jax_ignore|dno",
        skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
    },
    TeX: {
        equationNumbers: { autoNumber: "AMS" },
        noUndefined: { attributes: { mathcolor: "red", mathbackground: "#FFEEEE", mathsize: "90%" } },
        Macros: { href: "{}" }
    },
    messageStyle: "none"
    });
</script>
<!-- 给MathJax元素添加has-jax class -->
<script type="text/x-mathjax-config">
    MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax(), i;
        for(i=0; i < all.length; i += 1) {
            all[i].SourceElement().parentNode.className += ' has-jax';
        }
    });
</script>
<!-- 通过连接CDN加载MathJax的js代码 -->
<script type="text/javascript" async
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-MML-AM_CHTML">
</script>


    


<!-- hexo injector head_end start -->
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.12.0/dist/katex.min.css">

<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/hexo-math@4.0.0/dist/style.css">
<!-- hexo injector head_end end --><meta name="generator" content="Hexo 6.3.0"></head>

<body>
    <script>
        // this function is used to check current theme before page loaded.
        (() => {
            const currentTheme = window.localStorage && window.localStorage.getItem('theme') || '';
            const isDark = currentTheme === 'dark';
            const pagebody = document.getElementsByTagName('body')[0]
            if (isDark) {
                pagebody.classList.add('dark-theme');
                // mobile
                document.getElementById("mobile-toggle-theme").innerText = "· Dark"
            } else {
                pagebody.classList.remove('dark-theme');
                // mobile
                document.getElementById("mobile-toggle-theme").innerText = "· Light"
            }
        })();
    </script>

    <div class="wrapper">
        <header>
    <nav class="navbar">
        <div class="container">
            <div class="navbar-header header-logo"><a href="/">Liang Xu</a></div>
            <div class="menu navbar-right">
                
                    <a class="menu-item" href="/research">Research</a>
                
                    <a class="menu-item" href="/publication">Publications</a>
                
                    <a class="menu-item" href="/packages">Packages</a>
                
                    <a class="menu-item" href="/archives">Blog</a>
                
                    <a class="menu-item" href="/about">About</a>
                
                <input id="switch_default" type="checkbox" class="switch_default">
                <label for="switch_default" class="toggleBtn"></label>
            </div>
        </div>
    </nav>

    
    <nav class="navbar-mobile" id="nav-mobile">
        <div class="container">
            <div class="navbar-header">
                <div>
                    <a href="/">Liang Xu</a><a id="mobile-toggle-theme">·&nbsp;Light</a>
                </div>
                <div class="menu-toggle" onclick="mobileBtn()">&#9776; Menu</div>
            </div>
            <div class="menu" id="mobile-menu">
                
                    <a class="menu-item" href="/research">Research</a>
                
                    <a class="menu-item" href="/publication">Publications</a>
                
                    <a class="menu-item" href="/packages">Packages</a>
                
                    <a class="menu-item" href="/archives">Blog</a>
                
                    <a class="menu-item" href="/about">About</a>
                
            </div>
        </div>
    </nav>

</header>
<script>
    var mobileBtn = function f() {
        var toggleMenu = document.getElementsByClassName("menu-toggle")[0];
        var mobileMenu = document.getElementById("mobile-menu");
        if(toggleMenu.classList.contains("active")){
           toggleMenu.classList.remove("active")
            mobileMenu.classList.remove("active")
        }else{
            toggleMenu.classList.add("active")
            mobileMenu.classList.add("active")
        }
    }
</script>
            <div class="main">
                <div class="container">
    
    
        <div class="post-toc">
    <div class="tocbot-list">
    </div>
    <div class="tocbot-list-menu">
        <a class="tocbot-toc-expand" onclick="expand_toc()">Expand all</a>
        <a onclick="go_top()">Back to top</a>
        <a onclick="go_bottom()">Go to bottom</a>
    </div>
</div>

<script>
    var tocbot_timer;
    var DEPTH_MAX = 6; // 为 6 时展开所有
    var tocbot_default_config = {
        tocSelector: '.tocbot-list',
        contentSelector: '.post-content',
        headingSelector: 'h1, h2, h3, h4, h5',
        orderedList: false,
        scrollSmooth: true,
        onClick: extend_click,
    };

    function extend_click() {
        clearTimeout(tocbot_timer);
        tocbot_timer = setTimeout(function() {
            tocbot.refresh(obj_merge(tocbot_default_config, {
                hasInnerContainers: true
            }));
        }, 420); // 这个值是由 tocbot 源码里定义的 scrollSmoothDuration 得来的
    }

    document.ready(function() {
        tocbot.init(obj_merge(tocbot_default_config, {
            collapseDepth: 1
        }));
    });

    function expand_toc() {
        var b = document.querySelector('.tocbot-toc-expand');
        var expanded = b.getAttribute('data-expanded');
        expanded ? b.removeAttribute('data-expanded') : b.setAttribute('data-expanded', true);
        tocbot.refresh(obj_merge(tocbot_default_config, {
            collapseDepth: expanded ? 1 : DEPTH_MAX
        }));
        b.innerText = expanded ? 'Expand all' : 'Collapse all';
    }

    function go_top() {
        window.scrollTo(0, 0);
    }

    function go_bottom() {
        window.scrollTo(0, document.body.scrollHeight);
    }

    function obj_merge(target, source) {
        for (var item in source) {
            if (source.hasOwnProperty(item)) {
                target[item] = source[item];
            }
        }
        return target;
    }
</script>
    

    
    <article class="post-wrap">
        <header class="post-header">
            <h1 class="post-title">Get genomes, annotate genomes and analyze genomes </h1>
            
                <div class="post-meta">
                    
                        Author: <a itemprop="author" rel="author" href="/">Liang Xu</a>
                    

                    
                        <span class="post-time">
                        Date: <a href="#">April 1, 2023&nbsp;&nbsp;0:00:00</a>
                        </span>
                    
                    
                        <span class="post-category">
                    Category:
                            
                                <a href="/categories/Research/">Research</a>
                            
                                <a href="/categories/Research/bioinformatics/">bioinformatics</a>
                            
                        </span>
                    
                </div>
            
        </header>

        <div class="post-content">
            <p>A pipeline is created to estimate maximum growth rates of bacteria from a set of ASVs (fractions of genomes):</p>
<ol>
<li><p>Comparing ASVs with database to get the rough matches between genomes and ASVs; <strong>BLAST ASVs with database</strong> It is easy to feed ASVs fasta files to <a target="_blank" rel="noopener" href="https://blast.ncbi.nlm.nih.gov/Blast.cgi">BLAST</a> to get the matches;   </p>
</li>
<li><p>Fetch full genome sequences from <a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/">NCBI</a> by the genome ids;</p>
</li>
<li><p>Annotate genomes using <a target="_blank" rel="noopener" href="https://github.com/tseemann/prokka">Prokka</a>;</p>
</li>
<li><p>Using the annotated genomes to estimate the maximum growth rates of bacteria by <a target="_blank" rel="noopener" href="https://github.com/jlw-ecoevo/gRodon2">gRodon2</a>.</p>
</li>
</ol>
<p>The pipeline is tested on <a target="_blank" rel="noopener" href="https://www.hpc.caltech.edu/"><strong>Caltech HPC</strong></a>.</p>
<span id="more"></span>

<h2 id="Get-genome-fasta-files-from-a-tsv-file-generated-by-BLAST"><a href="#Get-genome-fasta-files-from-a-tsv-file-generated-by-BLAST" class="headerlink" title="Get genome fasta files from a tsv file generated by BLAST"></a>Get genome fasta files from a tsv file generated by BLAST</h2><p>This is a way to get genome data by genome id from a tsv file generated by BLAST. The tsv file contains information of the species id and the genomen hit ids. One example is provided in the repo. <code>sh</code> scripting is used to automate the downloading processes facilitated by the two tools <code>datasets</code> and <code>dataformat</code> from NCBI. The <strong>parallel fetching</strong> has been implemented. A job submission file is provided for the cluster usage.</p>
<h3 id="Instal-datasets-and-dataformat-from-NCBI"><a href="#Instal-datasets-and-dataformat-from-NCBI" class="headerlink" title="Instal datasets and dataformat from NCBI"></a>Instal <code>datasets</code> and <code>dataformat</code> from NCBI</h3><p>Follow the <a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/datasets/docs/v2/download-and-install/">link</a> to install these two tools. </p>
<h3 id="Run-getgenome-parallel-uniformdb-sh"><a href="#Run-getgenome-parallel-uniformdb-sh" class="headerlink" title="Run getgenome_parallel_uniformdb.sh"></a>Run <code>getgenome_parallel_uniformdb.sh</code></h3><p>Basically, what this script does is:</p>
<ol>
<li><p>Extract the species label, e.g. c497da3b39f30aceede6bec3b03cd100 and the genome id, e.g. GB_GCA_905618805.1 from each line in the tsv file. One species label may have several genome ids, also called hits;</p>
</li>
<li><p>Generate folders for each genome id; Different ASVs may have the same match with the same genome. The repeated matched genome is downloaed once;</p>
</li>
<li><p>Run <code>datasets</code> to download the genome data package. Make sure <code>datasets</code> is in the same folder as the <code>sh</code> script.</p>
</li>
</ol>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sh getgenome_parallel_uniformdb.sh</span><br></pre></td></tr></table></figure>

<p>The package contains a bunch to information. See <a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/datasets/docs/v2/reference-docs/data-packages/genome/">here</a> for details. A fasta file is included in the package;</p>
<ol start="4">
<li>Unzip the downloaded data.</li>
</ol>
<h3 id="Run-getgenome-parallel-sh-number-of-cores-filename-tsv"><a href="#Run-getgenome-parallel-sh-number-of-cores-filename-tsv" class="headerlink" title="Run getgenome_parallel.sh -number-of-cores -filename.tsv"></a>Run <code>getgenome_parallel.sh -number-of-cores -filename.tsv</code></h3><p>Basically, what this script does is as the same as <code>getgenome.sh</code>. The parallel processing is implemented via <a target="_blank" rel="noopener" href="https://www.gnu.org/software/parallel/">GNU <code>parallel</code></a>. Note that two arguments are needed to run this <code>sh</code> as follows, taking MacOS terminal as an example:</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sh getgenome_parallel.sh -number-of-cores -filename.tsv</span><br></pre></td></tr></table></figure>

<h3 id="Annotating-genomes-using-Prokka"><a href="#Annotating-genomes-using-Prokka" class="headerlink" title="Annotating genomes using Prokka"></a>Annotating genomes using <a target="_blank" rel="noopener" href="https://github.com/tseemann/prokka">Prokka</a></h3><p>The script is <code>prokka_genomes.sh</code></p>
<p>It reads the genomes sequences <code>.fasta</code> file and annotates and output a bunch of files. See  <a target="_blank" rel="noopener" href="https://github.com/tseemann/prokka">Prokka</a> for more details.</p>
<p>The output files will be stored in “prokkaoutput” folder. The genome annotations will be stored in the folder named after the genome. </p>
<h3 id="Caveats"><a href="#Caveats" class="headerlink" title="Caveats"></a>Caveats</h3><ol>
<li>Linux being tested. Running well on the Caltech cluster;</li>
</ol>
<h3 id="Future-work"><a href="#Future-work" class="headerlink" title="Future work"></a>Future work</h3><ol>
<li>Using <a target="_blank" rel="noopener" href="https://github.com/jlw-ecoevo/gRodon2">gRodon</a> to estimate the maximum growth rates.</li>
</ol>

        </div>

        
            <section class="post-copyright">
                
                    <p class="copyright-item">
                        <span>Author:</span>
                        <span>Liang Xu</span>
                    </p>
                
                
                    <p class="copyright-item">
                        <span>Permalink:</span>
                        <span><a href="http://xl0418.github.io/2023/04/01/2023-04-01-GetGenomeAnnotation/">http://xl0418.github.io/2023/04/01/2023-04-01-GetGenomeAnnotation/</a></span>
                    </p>
                
                
                    <p class="copyright-item">
                        <span>License:</span>
                        <span>Copyright (c) 2019 <a target="_blank" rel="noopener" href="http://creativecommons.org/licenses/by-nc/4.0/">CC-BY-NC-4.0</a> LICENSE</span>
                    </p>
                
                
                     <p class="copyright-item">
                         <span>Slogan:</span>
                         <span>Do you believe in <strong>DESTINY</strong>?</span>
                     </p>
                

            </section>
        
        <section class="post-tags">
            <div>
                <span>Tag(s):</span>
                <span class="tag">
                    
                    
                        <a href="/tags/sh/"># sh</a>
                    
                        
                </span>
            </div>
            <div>
                <a href="javascript:window.history.back();">back</a>
                <span>· </span>
                <a href="/">home</a>
            </div>
        </section>
        <section class="post-nav">
            
                <a class="prev" rel="prev" href="/2024/02/19/2024-02-19-bacdiversity/">Supplementary material to the talk ME53A-01 -- Ocean Science Meeting 2024</a>
            
            
            <a class="next" rel="next" href="/2023/03/20/2023-03-20-bioinfofresh/">The first week of getting into the realm of bioinformatics</a>
            
        </section>


    </article>
</div>

            </div>
            <footer id="footer" class="footer">
    <div class="copyright">
        <span>© Liang Xu | Powered by <a href="https://hexo.io" target="_blank">Hexo</a> & <a href="https://github.com/Siricee/hexo-theme-Chic" target="_blank">Chic</a></span>
    </div>
</footer>

    </div>
</body>

</html>