<!DOCTYPE HTML>
<html>
<head>
  <meta charset="utf-8">
  
  <title>Get genomes from NCBI and annotating genomes using Prokka | Liang Xu | This is Liang&#39;s blog for life and work archive.</title>

  
  <meta name="author" content="Liang Xu">
  

  
  <meta name="description" content="Fetching and annotating genomes pipelineA pipeline is established to go through a couple of things:

Comparing ASVs with database to get the rough mat">
  

  
  
  <meta name="keywords" content="sh">
  

  <meta id="viewport" name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, minimum-scale=1, user-scalable=no, minimal-ui">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">

  <meta property="og:title" content="Get genomes from NCBI and annotating genomes using Prokka"/>

  <meta property="og:site_name" content="Liang Xu"/>

  
  <meta property="og:image" content="/favicon.ico"/>
  

  <link href="/favicon.ico" rel="icon">
  <link rel="alternate" href="/atom.xml" title="Liang Xu" type="application/atom+xml">
  <link rel="stylesheet" href="/css/style.css" media="screen" type="text/css">
<!-- hexo injector head_end start -->
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.12.0/dist/katex.min.css">

<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/hexo-math@4.0.0/dist/style.css">
<!-- hexo injector head_end end --><meta name="generator" content="Hexo 6.3.0"></head>


<body>
<div class="blog">
  <div class="content">

    <header>
  <div class="site-branding">
    <h1 class="site-title">
      <a href="/">Liang Xu</a>
    </h1>
    <p class="site-description">This is Liang&#39;s blog for life and work archive.</p>
  </div>
  <nav class="site-navigation">
    <ul>
      
        <li><a href="/">Home</a></li>
      
        <li><a href="/about">About</a></li>
      
        <li><a href="/categories">Categories</a></li>
      
        <li><a href="/tags">Tags</a></li>
      
    </ul>
  </nav>
</header>

    <main class="site-main posts-loop">
    <article>

  
    
    <h3 class="article-title"><span>Get genomes from NCBI and annotating genomes using Prokka</span></h3>
    
  

  <div class="article-top-meta">
    <span class="posted-on">
      <a href="/2023/04/01/2023-04-01-GetGenomeAnnotation/" rel="bookmark">
        <time class="entry-date published" datetime="2023-04-01T07:00:00.000Z">
          2023-04-01
        </time>
      </a>
    </span>
  </div>


  

  <div class="article-content">
    <div class="entry">
      
        <h1 id="Fetching-and-annotating-genomes-pipeline"><a href="#Fetching-and-annotating-genomes-pipeline" class="headerlink" title="Fetching and annotating genomes pipeline"></a>Fetching and annotating genomes pipeline</h1><p>A pipeline is established to go through a couple of things:</p>
<ol>
<li><p>Comparing ASVs with database to get the rough matches between genomes and ASVs; <strong>Not included in this pipeline.</strong> It is easy to feed ASVs fasta files to <a target="_blank" rel="noopener" href="https://blast.ncbi.nlm.nih.gov/Blast.cgi">BLAST</a> to get the matches;   </p>
</li>
<li><p>Fetch the full genome sequences from <a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/">NCBI</a> by the genome ids;</p>
</li>
<li><p>Annotate the genomes using <a target="_blank" rel="noopener" href="https://github.com/tseemann/prokka">Prokka</a>;</p>
</li>
<li><p>Using the annotated genomes to estimate the maximum growth rates of bacteria by <a target="_blank" rel="noopener" href="https://github.com/jlw-ecoevo/gRodon2">gRodon2</a>.</p>
</li>
</ol>
<h1 id="Get-genome-fasta-files-from-a-tsv-file-generated-by-BLAST"><a href="#Get-genome-fasta-files-from-a-tsv-file-generated-by-BLAST" class="headerlink" title="Get genome fasta files from a tsv file generated by BLAST"></a>Get genome fasta files from a tsv file generated by BLAST</h1><p>This is a way to get genome data by genome id from a tsv file generated by BLAST. The tsv file contains information of the species id and the genomen hit ids. One example is provided in the repo. <code>sh</code> scripting is used to automate the downloading processes facilitated by the two tools <code>datasets</code> and <code>dataformat</code> from NCBI. The <strong>parallel fetching</strong> has been implemented. A job submission file is provided for the cluster usage.</p>
<h2 id="Instal-datasets-and-dataformat-from-NCBI"><a href="#Instal-datasets-and-dataformat-from-NCBI" class="headerlink" title="Instal datasets and dataformat from NCBI"></a>Instal <code>datasets</code> and <code>dataformat</code> from NCBI</h2><p>Follow the <a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/datasets/docs/v2/download-and-install/">link</a> to install these two tools. </p>
<h2 id="Run-getgenome-parallel-uniformdb-sh"><a href="#Run-getgenome-parallel-uniformdb-sh" class="headerlink" title="Run getgenome_parallel_uniformdb.sh"></a>Run <code>getgenome_parallel_uniformdb.sh</code></h2><p>Basically, what this script does is:</p>
<ol>
<li><p>Extract the species label, e.g. c497da3b39f30aceede6bec3b03cd100 and the genome id, e.g. GB_GCA_905618805.1 from each line in the tsv file. One species label may have several genome ids, also called hits;</p>
</li>
<li><p>Generate folders for each genome id; Different ASVs may have the same match with the same genome. The repeated matched genome is downloaed once;</p>
</li>
<li><p>Run <code>datasets</code> to download the genome data package. Make sure <code>datasets</code> is in the same folder as the <code>sh</code> script.</p>
</li>
</ol>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sh getgenome_parallel_uniformdb.sh</span><br></pre></td></tr></table></figure>

<p>The package contains a bunch to information. See <a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/datasets/docs/v2/reference-docs/data-packages/genome/">here</a> for details. A fasta file is included in the package;</p>
<ol start="4">
<li>Unzip the downloaded data.</li>
</ol>
<h2 id="Run-getgenome-parallel-sh-number-of-cores-filename-tsv"><a href="#Run-getgenome-parallel-sh-number-of-cores-filename-tsv" class="headerlink" title="Run getgenome_parallel.sh -number-of-cores -filename.tsv"></a>Run <code>getgenome_parallel.sh -number-of-cores -filename.tsv</code></h2><p>Basically, what this script does is as the same as <code>getgenome.sh</code>. The parallel processing is implemented via <a target="_blank" rel="noopener" href="https://www.gnu.org/software/parallel/">GNU <code>parallel</code></a>. Note that two arguments are needed to run this <code>sh</code> as follows, taking MacOS terminal as an example:</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sh getgenome_parallel.sh -number-of-cores -filename.tsv</span><br></pre></td></tr></table></figure>

<h2 id="Annotating-genomes-using-Prokka"><a href="#Annotating-genomes-using-Prokka" class="headerlink" title="Annotating genomes using Prokka"></a>Annotating genomes using <a target="_blank" rel="noopener" href="https://github.com/tseemann/prokka">Prokka</a></h2><p>The script is <code>prokka_genomes.sh</code></p>
<p>It reads the genomes sequences <code>.fasta</code> file and annotates and output a bunch of files. See  <a target="_blank" rel="noopener" href="https://github.com/tseemann/prokka">Prokka</a> for more details.</p>
<p>The output files will be stored in “prokkaoutput” folder. The genome annotations will be stored in the folder named after the genome. </p>
<h2 id="Caveats"><a href="#Caveats" class="headerlink" title="Caveats"></a>Caveats</h2><ol>
<li>Linux being tested. Running well on the Caltech cluster;</li>
</ol>
<h2 id="Future-work"><a href="#Future-work" class="headerlink" title="Future work"></a>Future work</h2><ol>
<li>Using <a target="_blank" rel="noopener" href="https://github.com/jlw-ecoevo/gRodon2">gRodon</a> to estimate the maximum growth rates.</li>
</ol>

      
    </div>

  </div>

  <div class="article-footer">
    <div class="article-meta pull-left">

    
      

    <span class="post-categories">
      <i class="icon-categories"></i>
        <a href="/categories/Research/">Research</a>, <a href="/categories/Research/bioinformatics/">bioinformatics</a>
    </span>
    

    
    

    <span class="post-tags">
      <i class="icon-tags"></i>
        <a href="/tags/sh/">sh</a>
    </span>
    

    </div>

    
  </div>
</article>

  









    </main>

    <footer class="site-footer">
  <p class="site-info">
    Proudly powered by <a href="https://hexo.io/" target="_blank">Hexo</a> and
    Theme by <a href="https://github.com/CodeDaraW/Hacker" target="_blank">Hacker</a>
    </br>
    
    &copy; 2023 Liang Xu
    
  </p>
</footer>
    
    
  </div>
</div>
</body>
</html>